### ISSUE TITLE
Resuming saving state

### ISSUE THREAD

**Issue by SirIzumi** [xinntao/BasicSR#64]
What is the right way to resume training?

Let's say I want to fine-tune PSNR model, and set
`"pretrain_model_G": "../experiments/pretrained_models/RRDB_PSNR_x4.pth"`
in train_ESRGAN.json

So next time if I want to continue, should I load the same .json file with `pretrain_model_G` set up to PSNR again, or should I remove this line and only tell it to do
`"resume_state": "../experiments/model/training_state/18000.state"` ?

I am afrain it can start fine-tuning PSNR from the scratch if I leave it.

**Comment by xinntao** [comment]
Let's say, after training you have got : 
```
experiments/model/training_state/18000.state
experiments/models/180000_G.pth
```
If you want to resume from this 180000 state, you need to  load the same .json file 
and you are supposed to set two options:
```
"pretrain_model_G": "../experiments/models/180000_G.pth"
"resume_state": "../experiments/model/training_state/18000.state"
```
In other words, you need to set these two options together and make sure they match each other (both for 180000th iteration).

I think it can be improved. Later I will improve it by only setting the resume_state.

**Comment by SirIzumi** [comment]
Very interesting! So we have to set it to, basically "Fine tune incomplete model G by using saved state with the same name" to properly continue.

This means me and many people on Game_Upscale subreddit did it wrong all along.

And another question if you don't mind: for me, descreasing `batch_size` on both Windows and Linux increases speed dramatically, setting it from default 16 to 1 makes first 200 iter go from ~4 minutes to ~50 seconds.

Wasn't it supposed to make it faster with larger batch sizes?

**Comment by SirIzumi** [comment]
Thank you! What is the right niter, lr_gain, lr_steps and other values for different batch_sizes options? I suppose default ones (500 000 iterations with 50000, 100000, 200000, 300000 steps) in train_ESRGAN.json are set up for batch_size 16.

**Comment by xinntao** [comment]
The current setting is for batch size 16. I have not tried other settings. Usually, with a larger batch size, we can set a larger learning rate and fewer iterations. 

BTW, if you want to fine-tune for other datasets, you can fine-tuen directly from ESRGAN, which takes less time. You could use the pretrained G and D ([download](https://drive.google.com/file/d/1V3WkFPLwFm0lNB-4Wk9HKhPXObTo9x2A/view?usp=sharing)). And set learning rate from 1e-4 (or smaller), with much less iterations.

**Comment by SirIzumi** [comment]
Wow, thanks a lot. This would be helpful.

I will test this ASAP, but I usually use PSNR model for initial fine-tuning, because my dataset (textures) is very different from DIV2K or any others and they work better with PSNR. But maybe fine-tuning ESRGAN will make it better

As far as I understand, I need to add `pretrain_model_D` line to .json file.

**Comment by xinntao** [comment]
This model is the same model as that in ESRGAN repo (the G), only providing the extra D. 
You can try fine-tuning from ESRGAN, maybe will be better and faster. 
Yes, you also need to add `pretrain_model_D`

**Comment by HengPiPi** [comment]
> "pretrain_model_G": "../experiments/models/180000_G.pth" "resume_state": "../experiments/model/training_state/18000.state"

if I set

> "pretrain_model_G": null

> "resume_state":"../experiments/model/training_state/18000.state"
Is it the same as starting again?

**Comment by xinntao** [comment]
@HengPiPi 
I think you are right.

(@SirIzumi, sorry that I made mistakes in the previous response)  I have checked the codes again.  If you want to resume, you can just set the `resume_state` option.
The codes will automatically look for the corresponding `pretrain_model_G`.
The codes are here:
https://github.com/xinntao/BasicSR/blob/4f3889667b5bcab16711c69d75db9754945ff212/codes/options/options.py#L106-L120

**Comment by xkunglu** [comment]
Thank you for an excellently written code base.
Im here because Im trying to continue training using datasets with different statistics than the ones you trained on.Most yaml parameters were straightforwards to decipher by following your code, but i - like sirzizumi was having issues with updating the resume point. As a general point, comments on the .yaml parameters would be useful.

**Comment by xinntao** [comment]
@xkunglu Thanks. I will write a wiki later.

**Comment by wolfam0108** [comment]
Thank!
After deleting the lines, training began.

How can I interrupt my workout and resume?
Where are the training_state files stored or where can I specify a directory for them?

**Comment by xinntao** [comment]
@wolfam0108 The training_state files are stored in your experiment folder. 

For resuming, you should use the same exp-name as the interrupt one and modify the `resume_state` field in the yml option file to pointing to the  `training_state file`~

**Comment by vova0108** [comment]
Do training_state files appear immediately after starting a workout?
I have a folder created with the name of the project, but the training_state folder is empty.

**Comment by xinntao** [comment]
@vova0108 
The training state will be saved when saving models, which could be set by the key `save_checkpoint_freq: !!float 5e3` in the yml option file.

**Comment by vova0108** [comment]
Is it enough !!float 5e3? Or I need to change this? Is there an example?

**Comment by xinntao** [comment]
@vova0108 
`!!float 5e3` is an example, which means that after 5000 iterations, the program will save the checkpoint and training state. 

You can change it accordingly.
